Start of Metal Replay resource creation

Zorro666 opened the pull request:
## Description

Metal Replay resource creation for:
```
MTLCommandQueue
MTLLibrary
MTLBuffer
MTLRenderPipelineState
MTLTexture
MTLFunction
```

`MTLCommandBuffer`, `MTLRenderCommandEncoder`, `MTLBlitCommandEncoder`resource creation will be in a future PR.

New `WrappedMTLDevice` Replay resource creation helper methods:
```
void AddResource(ResourceId id, ResourceType type, const char *defaultNamePrefix);
void DerivedResource(ResourceId parentLive, ResourceId child);
template <typename MetalType>
void DerivedResource(MetalType parent, ResourceId child);
```

Start of `MetalReplay` class implementation to support replay resource creation:
`ResourceDescription & MetalReplay::GetResourceDesc(ResourceId id);`

The creation of the `MetalReplay` instance will be in a future PR.

Zorro666 committed:
Metal Replay resource creation helper methods

Added WrappedMTLDevice
void AddResource(ResourceId id, ResourceType type, const char *defaultNamePrefix);
void DerivedResource(ResourceId parentLive, ResourceId child);
template <typename MetalType> void DerivedResource(MetalType parent, ResourceId child);

Start of MetalReplay implementation to support replay resource creation:
ResourceDescription &GetResourceDesc(ResourceId id);

MetalReplay instance is not created in this commit

Zorro666 committed:
Metal Replay resource creation of a few base types

MTLCommandQueue
MTLLibrary
MTLBuffer
MTLRenderPipelineState
MTLTexture
MTLFunction

Zorro666 commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -179,7 +179,13 @@ bool WrappedMTLDevice::Serialise_newCommandQueue(SerialiserType &ser, WrappedMTL
   if(IsReplayingAndReading())
   {
-    // TODO: implement RD MTL replay
+    MTL::CommandQueue *realMTLCommandQueue = Unwrap(this)->newCommandQueue();
```
A question for all the code which is creating the live resources.
What code should be written or asserts added to handle or guard if the live resource is created multiple times? 
Currently, this resource creation would leak a live resource if it is called multiple times with the same resource ID.

baldurk replied:
That should not be possible, since IDs are monotonically increasing (can't be re-used) and creation chunks should only be created once. That would indicate some bug at capture time that caused a duplicate chunk or something like that.

The resource manager already has an error message if a live ID is registered twice, which would be enough to not have this lost silently.

Zorro666 commented on renderdoc/driver/metal/metal_replay.h:
```diff
@@ -0,0 +1,40 @@
+  ResourceDescription &GetResourceDesc(ResourceId id);
+
+private:
+  WrappedMTLDevice *m_pDriver = NULL;
```
This could be a reference if that makes more sense or is a preferred style

baldurk replied:
I think that's up to you, I've tended to keep it as a pointer usually but you're entirely right that it could be a reference.

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -430,6 +480,20 @@ bool WrappedMTLDevice::Serialise_newTextureWithDescriptor(SerialiserType &ser,
+    // Ensure the created textures can be read by a shader
+    // Metal driver will treat TextureUsageUnknown as all options
+    if(descriptor.usage != MTL::TextureUsageUnknown)
+      descriptor.usage = (MTL::TextureUsage)(descriptor.usage | MTL::TextureUsageShaderRead);
```
Just to check - is there any possible side-effect of making this change in metal, or is it entirely independent?

On D3D11 e.g. we could freely set these bits only on replay and it wouldn't change anything. On Vulkan at least we need to set these bits during capture too to ensure that queried memory properties are respected, as they can change with different usage flags. I assume metal falls into the former camp, but it's worth making sure.

baldurk reviewed (approved):
This looks good to me. I've left one query about usage flags that I'd like to clarify but otherwise it seems fine to merge,

Zorro666 committed:
Set MTL::TextureUsageShaderRead during capture

To be consistent with replay created textures set MTL::TextureUsageShaderRead on textures created during capture. This override is not serialized.

PR Feedback
